### PR TITLE
statistics: improve query locked table API memory usage

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/46351

Problem Summary:

### What is changed and how it works?

Only return locked tables to reduce memory usage.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -181,32 +181,38 @@ func filterAndCollectTasks(tasks []*analyzeTask, statsHandle *handle.Handle, inf
 		tids                = make([]int64, 0, len(tasks))
 		// taskMap is used to collect the tasks that belong to the same table.
 		// In stats v1, analyze for each index is a single task, and they have the same table id.
-		taskMap = make(map[int64][]*analyzeTask, len(tasks))
+		tidMap = make(map[int64]struct{}, len(tasks))
```
Comment (time-and-fate) [inline executor/analyze.go:184]: Please also update the comments above.

File: executor/analyze.go [hunk 2]
```diff
@@ -181,32 +181,38 @@ func filterAndCollectTasks(tasks []*analyzeTask, statsHandle *handle.Handle, inf
...
+	// Check the locked tables in one transaction.
 	for _, task := range tasks {
 		tableID := getTableIDFromTask(task)
 		tids = append(tids, tableID)
-		taskMap[tableID] = append(taskMap[tableID], task)
 	}
-
-	// Check the locked tables in one transaction.
-	lockedStatuses, err := statsHandle.QueryTablesLockedStatuses(tids...)
+	lcokedTables, err := statsHandle.GetLockedTables(tids...)
```
Comment (time-and-fate) [inline executor/analyze.go:192]: typo

File: executor/analyze.go [hunk 3]
```diff
@@ -181,32 +181,38 @@ func filterAndCollectTasks(tasks []*analyzeTask, statsHandle *handle.Handle, inf
...
-			tbl, ok := infoSchema.TableByID(tid)
-			if !ok {
-				logutil.BgLogger().Warn("Unknown table ID in analyze task", zap.Int64("tid", tid))
+			filteredTasks = append(filteredTasks, task)
+		}
+		if _, ok := tidMap[tableID]; !ok {
+			if isLocked {
+				tbl, ok := infoSchema.TableByID(tableID)
+				if !ok {
+					// Ignore this table because it may have been dropped.
```
Comment (time-and-fate) [inline executor/analyze.go:207]: Looks like this line is mistakenly added back.

File: executor/analyze.go (hunk 1 above)
Comment (hi-rustin) [reply executor/analyze.go:184]: Oops. Updated.

File: executor/analyze.go (hunk 2 above)
Comment (hi-rustin) [reply executor/analyze.go]: Fixed.

File: executor/analyze.go (hunk 3 above)
Comment (hi-rustin) [reply executor/analyze.go]: Removed.

**Review by hawkingrei — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**